File: lang/binary_tree.py
import time


class Tree:
    def __init__(self, item, depth):
        self.item = item
        self.depth = depth
        if depth > 0:
            self.left = Tree(2 * item - 1, depth - 1)
            self.right = Tree(2 * item, depth - 1)
        else:
            self.left = None
            self.right = None

    def check(self):
        if self.left:
            return self.item + self.left.check() - self.right.check()
        else:
            return self.item


min_depth = 4
max_depth = 14
stretch_depth = max_depth + 1

start = time.time()
print("stretch tree of depth " + repr(stretch_depth))
print("check: " + repr(Tree(0, stretch_depth).check()))

long_lived_tree = Tree(0, max_depth)
iterations = 2 ** max_depth

depth = min_depth
while depth < stretch_depth:
    check = 0
    for i in range(1, iterations + 1):
        check += Tree(i, depth).check() + Tree(-i, depth).check()

    print("num tree: " + repr(iterations * 2))
    print("depth: " + repr(depth))
    print("check: " + repr(check))

    iterations //= 4
    depth += 2

print("long lived tree of depth: " + repr(max_depth))
print("check: " + repr(long_lived_tree.check()))
print("elapsed: " + repr(time.time() - start))

File: lang/binary_tree.js
class Tree {
    constructor(item, depth) {
        this.item = item;
        this.depth = depth;
        if (depth > 0) {
            this.left = new Tree(2 * item - 1, depth - 1);
            this.right = new Tree(2 * item, depth - 1);
        } else {
            this.left = this.right = null;
        }
    }

    check() {
        if (this.left !== null) {
            return this.item + this.left.check() - this.right.check();
        } else {
            return this.item;
        }
    }
}

const min_depth = 4;
const max_depth = 14;
const stretch_depth = max_depth + 1;

const start = Date.now() / 1000;

console.log('streitch tree of depth ' + stretch_depth);
console.log('check: ' + (new Tree(0, stretch_depth)).check());

const long_lived_tree = new Tree(0, max_depth);
let iterations = 2 ** max_depth;

let depth = min_depth;
while (depth < stretch_depth) {
    let check = 0;
    for (let i = 1; i <= iterations; i += 1) {
        check += (new Tree(i, depth)).check() + (new Tree(-i, depth)).check();
    }

    console.log('num tree: ' + iterations * 2);
    console.log('depth: ' + depth);
    console.log('check: ' + check);

    iterations /= 4;
    depth += 2;
}

console.log('long lived tree of depth:' + max_depth);
console.log('check: ' + long_lived_tree.check());
console.log(Date.now() / 1000 - start);